### PR TITLE
Added some extra tests to subscriptions_spec.rb

### PR DESCRIPTION
**Pull request by leamha**
#### What? Why?
I added some more tests in this file to help with the robustness of the test suite
Addresses #7454

There was a request to add more coverage to the tests for subscription and I believe I was able to do that. I added one case where the bill_address_phone is reset to nothing after being set to '0400 123 456' and check that the page updates with one can't be blank message, then set the phone back to 0400 123 456. I noticed that after copying to the shipping address only a few fields were tested despite there being more fields listed in the _address.html.haml file so I tested that the extra fields were also written in. Finally I added a test where the first item a customer adds a product and they set the quantity to 0, to make sure the price is also zero at that point. I was going to add a test for 'Please add at least one product' but I was not sure if that would cause the test case to fail since technically there was a product added with zero quantity. However, if that message should appear with a quantity of zero for a product I believe it could be added after line 255 in the file.

#### What should we test?
Since this is meant for coverage determine if the coverage from this test increase. Additionally, when shipping, make sure all fields of the shipping address are copied over if the user clicks on copy and I suppose determine if a product with a quantity of zero should warrant a 'Please add at least one product' message as I couldn't determine the functionality for that myself (I am new to this project).

#### Release notes
Added extra (simple) tests to the subscriptions_spec.rb

### CODE REVIEW

**Review by andrewpbrett — APPROVED** [review]
Thanks for this first contribution @leamha! 

I think this looks like a good addition to merge. It looks like it will increase test coverage, according to Codecov. 

As @filipefurtad0 mentioned, this is a particularly tricky section of the codebase that could really use a lot of refactoring. So I think we'll want to keep the issue open even once this is merged.

Let us know if you need help finding another good issue to work on!

**Comment by leamha** [comment]
> Thanks for this first contribution @leamha!
> 
> I think this looks like a good addition to merge. It looks like it will increase test coverage, according to Codecov.
> 
> As @filipefurtad0 mentioned, this is a particularly tricky section of the codebase that could really use a lot of refactoring. So I think we'll want to keep the issue open even once this is merged.
> 
> Let us know if you need help finding another good issue to work on!

Thank you very much @andrewpbrett , I expected after contributing to this as best as I could the issue would remain open. This is the first time I've contributed to a code base so as a quick question am I expected to merge the issue myself since it has been approved or will that be done by someone else? I just want to make sure I follow the correct process this community has with merging a contribution but I also don't want to step on any toes if I were to do it myself. Thanks for all of your help!

**Comment by andrewpbrett** [comment]
Hi @leamha,

> am I expected to merge the issue myself

No, there needs to be a second review done before it can be merged. Normally there would also be a manual testing/verification step, but in this case, since the only changes were to the specs, that's unnecessary. I'll add the pr-no-test label to reflect that. 

You can read more about the process on our wiki page: 
https://github.com/openfoodfoundation/openfoodnetwork/wiki/The-process-of-review%2C-test%2C-merge-and-deploy

**Comment by leamha** [comment]
> I did just notice, however, that there's a failing spec:
> 
> ```
> Subscriptions as an enterprise user creating a new subscription passes the smoke test
>      Failure/Error: expect(element.value).to eq(opts[:with]) if element && opts.key?(:with)
> 
>        expected: "Australia"
>             got: "1"
>        (compared using ==)
>      # ./spec/support/request/web_helper.rb:28:in `have_input'
>      # ./spec/features/admin/subscriptions_spec.rb:260:in `block (4 levels) in <top (required)>'
> ```
> 
> Could you look into what changed there and why it's failing?
 I looked into it and removed the lines that I believe would cause the issue and re-pushed!

**Comment by andrewpbrett** [comment]
Thanks @leamha! It looks like there's now a different failure: `./spec/features/admin/subscriptions_spec.rb:268`. Does that test pass locally for you? You can do `bundle exec rspec ./spec/features/admin/subscriptions_spec.rb:268` to check.

**Comment by leamha** [comment]
> Thanks @leamha! It looks like there's now a different failure: `./spec/features/admin/subscriptions_spec.rb:268`. Does that test pass locally for you? You can do `bundle exec rspec ./spec/features/admin/subscriptions_spec.rb:268` to check.

It was passing locally for me, however I'll be sure to look into it again and see if I can figure it out. I currently have exams for school so it may be a few days before I can get back on this issue.

File: spec/features/admin/subscriptions_spec.rb
Comment (filipefurtad0) [inline spec/features/admin/subscriptions_spec.rb]: I see your second commit removes these assertions. But I think it might be a good idea to check them as you did initially, only using a different matcher: `ship_address_country_id` and `ship_address_state_id` are selectors and input not fields, so I think `have_input` is triggering the error. One idea could be trying `have_selector` instead:
`expect(page).to have_selector '<dropdown_identifyer>', text: '<dropdown_option>'`

File: spec/features/admin/subscriptions_spec.rb
Comment (filipefurtad0) [inline spec/features/admin/subscriptions_spec.rb]: I see this second commit removes these assertions. But I think it might be a good idea to check them as you did initially, only using a different matcher: `ship_address_country_id` and `ship_address_state_id` are selectors and input not fields, so I think `have_input` is triggering the error. One idea could be trying `have_selector` instead:
`expect(page).to have_selector '<dropdown_identifyer>', text: '<dropdown_option>'`

**Review by filipefurtad0 — COMMENTED** [review]
Thank you for your contribution @leamha! 

I totally second you @andrewpbrett: I think this file needs serious refactoring. I think the `it` block starting on line 187 `it "passes the smoke test" do` contains several examples which actually correspond to very different scenarios. I think this might lead to the whole block to fail, making it harder to track down what is causing failure.

We should probably split this block in a ["into one-group,-one-example"](https://relishapp.com/rspec/rspec-core/v/3-8/docs/example-groups/basic-structure-describe-it#one-group,-one-example) fashion. We can do this in a seperate PR, I'll open an issue to address this.

File: spec/features/admin/subscriptions_spec.rb
Comment (Matt-Yorkley) [inline spec/features/admin/subscriptions_spec.rb:230]: We tend to prefer using double quotes as a general style thing, and in this case it would mean you don't need to escape the single quote:
```diff
-- 'can\'t be blank'
++ "can't be blank"
```

File: spec/features/admin/subscriptions_spec.rb
Comment (filipefurtad0) [inline spec/features/admin/subscriptions_spec.rb:250]: The build is failing due to these lines, in which a variant with zero quantity is being added. If one adds `byebug` and removes the `headless` from `spec_helper.rb` (see our [wiki for some tips on debugging](https://github.com/openfoodfoundation/openfoodnetwork/wiki/Testing-and-Rspec-Tips#debugging--tests)) we can check what's happening on the browser:

image

Above we can see that the line item is not being added to the order, because the UI is not accepting zero as a value; opening the console we can check that this input field has a condition of `min=1`, see:
Peek 2021-05-05 11-50

So, I think this test case is useful but it is not valid right now, because it expects a line item to be added. We could consider removing the `within` block - it's causing this failure on the build:

`        expected to find css "#subscription-line-items .item" 1 time but there were no matches`

and replace it with an assertion expecting no line item to be added, when quantity is set to zero.

**Comment by andrewpbrett** [comment]
We're going to close this since we're about to pick up some work on refactoring subscriptions and this has gotten a bit stale. Thank you for the contribution @leamha!

**Pull request closed without merging**
